### ISSUE TITLE
UV_THREADPOOL_SIZE setting for Oracle / libuv issue

### ISSUE THREAD

**Issue by cjbj** [knex/knex#4076]
# Environment 

Knex version: GitHub master
Database + version: Oracle
OS: all

# Bug
An attempt to set the libuv UV_THREADPOOL_SIZE in the code for the Oracle dialect is in https://github.com/knex/knex/blob/80355e48509857fb4941f65f4b611973a9ee99d5/lib/dialects/oracledb/index.js#L18-L26  
This setting (i) needs to be made before the apps threadpool is started, so this function looks to be too late. (ii) won't work on Windows for some reason known to Node.js or libuv.  The environment variable needs to be set before Node.js starts, eg. in a package.json 'start' script.

Also I could quibble over https://github.com/knex/knex/blob/7cfb78d16aa9dcd635c9ae1862837ccd4f507a21/CHANGELOG.md#L869 which refers the UV_THREADPOOL_SIZE as "Oracle's".  It is actually Node.js's or technically libuvs.

I don't have a good solution - other than documentation.

**Comment by cjbj** [comment]
Tagging @atiertant per the issue template.

**Comment by kibertoad** [comment]
@cjbj So what happens currently if user correctly sets this value in his start script, and then we override it within our code? Does that break things? Also is there any documentation for recommended values for using Oracle that we could refer to from our documentation?

**Comment by cjbj** [comment]
As far as I can tell from the [libuv code](https://github.com/libuv/libuv/blob/72fe3543feb23ae555e08628b70a3fae4da5706c/src/threadpool.c#L194), `UV_THREADPOOL_SIZE` is only ever checked once during thread pool initialization, so overriding it inside Knex won't have an effect unless it is done before the thread pool starts.  One problem with overriding it is that it makes it hard to see what size was actually used, since there is no way (?) to get the actual pool size from Node.js. The env var is the next best thing for logging the size but can be 'wrong' if the var was set after the pool started.

The node-oracledb doc about this is at [Connections, Threads, and Parallelism](https://oracle.github.io/node-oracledb/doc/api.html#numberofthreads).

**Comment by kibertoad** [comment]
@cjbj Thank! I'll update the docs.

**Comment by aidenfoxx** [comment]
After digging into this issue I have some unfortunate findings. Firstly that the code which sets `UV_THREADPOOL_SIZE` doesn't work, and will actually set a `UV_THREADPOOL_SIZE` of NaN since `this.driver.poolMax` is undefined. I assume the correct value should be `this.pool.max`.

However, this doesn't really matter since `UV_THREADPOOL_SIZE` is ignored unless its set before execution. At least on Ubuntu 20.04.

It seems like the only thing we really can do is make it clear that Oracle users should set their `UV_THREADPOOL_SIZE` appropriately.

**Comment by atiertant** [comment]
@cjbj @kibertoad in our tests, if `UV_THREADPOOL_SIZE` wasn't set correctly, some transaction never finish without any error.
maybe driver has change... 
see https://github.com/oracle/node-oracledb/issues/395

**Comment by aidenfoxx** [comment]
@atiertant I also experienced this. And the suggestion in that thread about releasing the connection doesn't help.

For better debugging I replaced the tarn pool with the default oracle pool for the oracledb dialect, with explicit connection releasing whenever anything finishes. The connections would still lock up as soon as the pool exceeded the `UV_THREADPOOL_SIZE`.

Feels to me like this is an issue with `node-oracledb`, and the best we can do is throw a warning to the user when the pool size is greater than the `UV_THREADPOOL_SIZE`, since changing `UV_THREADPOOL_SIZE` doesn't work.

**Comment by cjbj** [comment]
@aidenfoxx @atiertant You definitely need UV_THREADPOOL_SIZE set correctly - at the right place and right value.  Some users have (had) it quite high to cope with code that tries constructs like `Parallel.all()` on a single connection.  In node-oracledb 5.2 the [execution queue enhancement](https://github.com/oracle/node-oracledb/commit/6700e39f645c956f3d9a503e0eaef80340598177) can help reduce the need for threads, but you still need one per connection plus extra for non-DB work.  If you are playing with an implementation that uses Oracle's pool, make sure you are closing connections even on error conditions 0 check the [pool stats](https://oracle.github.io/node-oracledb/doc/api.html#connpoolmonitor) to see what's going on.

**Comment by cjbj** [comment]
@atiertant I think the UV_THREADPOOL_SIZE setting code should be removed. It doesn't do anything because it set after the thread pool has already started.

```
@@ -22,11 +22,6 @@
 class Client_Oracledb extends Client_Oracle {
   constructor(config) {
     super(config);
-    if (this.driver) {
-      process.env.UV_THREADPOOL_SIZE = process.env.UV_THREADPOOL_SIZE || 1;
-      process.env.UV_THREADPOOL_SIZE =
-        parseInt(process.env.UV_THREADPOOL_SIZE) + this.driver.poolMax;
-    }
   }
 
   _driver() {
```

**Comment by aidenfoxx** [comment]
I agree that the code should be removed. I would prefer instead to limit the `poolMax` value based `UV_THREADPOOL_SIZE`, and log a warning to the user that their threadpool size is incompatible with their requested pool size.

Happy to submit a PR for this if @kibertoad agrees?

**Comment by mercmobily** [comment]
Fast forward to 2025 (nearly 2026!)...

The current code in the Oracle dialect attempts to adjust `UV_THREADPOOL_SIZE` at runtime, which is ineffective (too late) and doesn’t work on Windows. This should be removed, and we should document correct sizing and constraints instead. This also ties into broader pooling meta work (#6241)

Code evidence:

- lib/dialects/oracledb/index.js:21 — constructor sets `process.env.UV_THREADPOOL_SIZE = process.env.UV_THREADPOOL_SIZE || 1; process.env.UV_THREADPOOL_SIZE = parseInt(process.env.UV_THREADPOOL_SIZE) + this.driver.poolMax;` — too late to affect the Node/libuv pool

Recommended path:

- Remove the runtime mutation of `UV_THREADPOOL_SIZE` from the dialect.
- Add docs recommending setting `UV_THREADPOOL_SIZE` before Node starts (e.g. via npm scripts) and guidance for sizing relative to pool size (at least `poolMax` plus some extra for non‑DB work). Link to node‑oracledb docs “Connections, Threads, and Parallelism”.
- Optionally, warn at client init if `poolMax` exceeds the detected threadpool size.

Thanks @cjbj, @kibertoad, and @atiertant for the input in the discussion; these changes align with it.
